File: kudos-ao3-reader/Features/Library/ReadingStatistics.swift
import Foundation

/// Local-only reading insights derived from the Library's `SavedWork` records.
/// The app does not track sessions or per-word progress, so "words read" counts
/// only finished works whose AO3 word count is known.
struct ReadingStatistics {
    struct FandomCount: Identifiable, Equatable {
        let name: String
        let count: Int

        var id: String {
            name
        }
    }

    let totalWorks: Int
    let startedWorks: Int
    let finishedWorks: Int
    let inProgressWorks: Int
    let wordsRead: Int
    let openedLast7Days: Int
    let openedLast30Days: Int
    let latestReadDate: Date?
    let topFandoms: [FandomCount]

    var completionRate: Double {
        guard startedWorks > 0 else { return 0 }
        return Double(finishedWorks) / Double(startedWorks)
    }

    init(
        works: [SavedWork],
        now: Date = Date(),
        calendar: Calendar = .current
    ) {
        var started: [SavedWork] = []
        for work in works where Self.hasStarted(work) {
            started.append(work)
        }
        let finished = works.filter(\.isFinished)
        let inProgress = started.filter { !$0.isFinished }

        totalWorks = works.count
        startedWorks = started.count
        finishedWorks = finished.count
        inProgressWorks = inProgress.count
        wordsRead = finished.reduce(0) { total, work in
            total + max(0, work.wordCount)
        }
        latestReadDate = works.compactMap(\.lastReadDate).max()

        let startOfToday = calendar.startOfDay(for: now)
        let sevenDayStart = calendar.date(byAdding: .day, value: -6, to: startOfToday)
            ?? startOfToday
        let thirtyDayStart = calendar.date(byAdding: .day, value: -29, to: startOfToday)
            ?? startOfToday
        openedLast7Days = Self.countOpened(works, since: sevenDayStart, through: now)
        openedLast30Days = Self.countOpened(works, since: thirtyDayStart, through: now)
        topFandoms = Self.fandomCounts(in: started)
    }

    /// Finished counts as started even when the progress fields were reset. Defers
    /// to the model's canonical `hasStartedReading` — a private re-listing of its
    /// fields here once drifted (it missed the Readium reader's locator, undercounting
    /// works read only on iOS).
    private static func hasStarted(_ work: SavedWork) -> Bool {
        work.isFinished || work.hasStartedReading
    }

    private static func countOpened(
        _ works: [SavedWork],
        since start: Date,
        through end: Date
    ) -> Int {
        works.reduce(into: 0) { count, work in
            guard let date = work.lastReadDate, date >= start, date <= end else { return }
            count += 1
        }
    }

    private static func fandomCounts(in works: [SavedWork]) -> [FandomCount] {
        var counts: [String: Int] = [:]
        for work in works {
            let uniqueFandoms = Set(work.workFandoms.compactMap { raw -> String? in
                let fandom = raw.trimmingCharacters(in: .whitespacesAndNewlines)
                return fandom.isEmpty ? nil : fandom
            })
            for fandom in uniqueFandoms {
                counts[fandom, default: 0] += 1
            }
        }
        return counts
            .map(FandomCount.init(name:count:))
            .sorted {
                if $0.count != $1.count { return $0.count > $1.count }
                return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
            }
    }
}

File: android/app/src/main/java/io/github/cidy02/kudos/library/ReadingStatistics.kt
package io.github.cidy02.kudos.library

import io.github.cidy02.kudos.core.model.SavedWork
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZonedDateTime

/**
 * Local-only reading insights derived from the Library's [SavedWork] records.
 *
 * The app does not track sessions or per-word progress, so "words read" counts
 * only finished works whose AO3 word count is known (non-negative; unknown → 0).
 *
 * Mirrors Apple `ReadingStatistics`.
 */
data class ReadingStatistics(
    val totalWorks: Int,
    val startedWorks: Int,
    val finishedWorks: Int,
    val inProgressWorks: Int,
    val wordsRead: Int,
    val openedLast7Days: Int,
    val openedLast30Days: Int,
    val latestReadDate: Instant?,
    val topFandoms: List<FandomCount>
) {
    data class FandomCount(
        val name: String,
        val count: Int
    )

    /** Finished / started; 0 when nothing has been started. */
    val completionRate: Double
        get() = if (startedWorks > 0) {
            finishedWorks.toDouble() / startedWorks.toDouble()
        } else {
            0.0
        }

    companion object {
        /**
         * Compute statistics from library works.
         *
         * @param works All saved (non-deleted) works.
         * @param now Reference instant (tests inject a fixed clock).
         * @param zone Calendar zone for day-window boundaries (defaults to system).
         */
        fun from(
            works: List<SavedWork>,
            now: Instant = Instant.now(),
            zone: ZoneId = ZoneId.systemDefault()
        ): ReadingStatistics {
            val started = works.filter { hasStarted(it) }
            val finished = works.filter { it.isFinished }
            val inProgress = started.filter { !it.isFinished }

            val wordsRead = finished.sumOf { work -> maxOf(0, work.wordCount) }
            val latestReadDate = works.mapNotNull { it.lastReadDate }.maxOrNull()

            val zonedNow = ZonedDateTime.ofInstant(now, zone)
            val startOfToday = zonedNow.toLocalDate()
            // Inclusive windows ending at `now`: past 7 calendar days = today + 6 prior.
            val sevenDayStart = startOfDay(startOfToday.minusDays(6), zone)
            val thirtyDayStart = startOfDay(startOfToday.minusDays(29), zone)

            return ReadingStatistics(
                totalWorks = works.size,
                startedWorks = started.size,
                finishedWorks = finished.size,
                inProgressWorks = inProgress.size,
                wordsRead = wordsRead,
                openedLast7Days = countOpened(works, since = sevenDayStart, through = now),
                openedLast30Days = countOpened(works, since = thirtyDayStart, through = now),
                latestReadDate = latestReadDate,
                topFandoms = fandomCounts(started)
            )
        }

        /**
         * A work counts as started if the user finished it, opened it, or has
         * non-zero progress. Matches Apple `ReadingStatistics.hasStarted`.
         */
        fun hasStarted(work: SavedWork): Boolean {
            return work.isFinished ||
                work.lastReadDate != null ||
                work.lastSpineIndex > 0 ||
                work.lastScrollFraction > 0.0
        }

        private fun countOpened(
            works: List<SavedWork>,
            since: Instant,
            through: Instant
        ): Int {
            return works.count { work ->
                val date = work.lastReadDate ?: return@count false
                !date.isBefore(since) && !date.isAfter(through)
            }
        }

        /**
         * Count each distinct fandom once per started work; sort by count desc,
         * then name case-insensitive ascending.
         */
        private fun fandomCounts(works: List<SavedWork>): List<FandomCount> {
            val counts = linkedMapOf<String, Int>()
            for (work in works) {
                val unique = work.workFandoms
                    .map { it.trim() }
                    .filter { it.isNotEmpty() }
                    .toSet()
                for (fandom in unique) {
                    counts[fandom] = (counts[fandom] ?: 0) + 1
                }
            }
            return counts.entries
                .map { (name, count) -> FandomCount(name = name, count = count) }
                .sortedWith(
                    compareByDescending<FandomCount> { it.count }
                        .thenBy(String.CASE_INSENSITIVE_ORDER) { it.name }
                )
        }

        private fun startOfDay(date: LocalDate, zone: ZoneId): Instant {
            return date.atStartOfDay(zone).toInstant()
        }
    }
}
